src/chainstate/stacks/boot/bns.clar

```diff
@@ -0,0 +1,973 @@
+  (or 
+    (is-digit char)
+    (is-special-char char)))
+
+(define-private (has-vowels-chars (name (buff 16)))
+  (> (len (filter is-vowel name)) u0))
+
+(define-private (has-nonalpha-chars (name (buff 16)))
+  (> (len (filter is-nonalpha name)) u0))
+
+(define-private (has-invalid-chars (name (buff 19)))
+  (< (len (filter is-char-valid name)) (len name)))
+
+(define-private (compute-name-price (name (buff 16))
+                                    (price-function (tuple (buckets (list 16 uint)) (base uint) (coeff uint) (nonalpha-discount uint) (no-vowel-discount uint))))
+  (let (
+    (exponent (get-exp-at-index (get buckets price-function) (min u15 (- (len name) u1))))
+    (no-vowel-discount (if (not (has-vowels-chars name)) (get no-vowel-discount price-function) u1))
+    (nonalpha-discount (if (has-nonalpha-chars name) (get nonalpha-discount price-function) u1)))
+    (*
+      (/
+        (*
+          (get coeff price-function)
+          (pow (get base price-function) exponent))
+        (max nonalpha-discount no-vowel-discount))
+      u10))) ;; 10 = name_cost (100) * "old_price_multiplier" (0.1) - todo(ludo): sort this out.
+
+(define-private (is-name-lease-expired? (namespace (buff 19)) (name (buff 16)))
+  (let (
+    (namespace-props (unwrap! 
+      (map-get? namespaces ((namespace namespace))) 
+      (err ERR_NAMESPACE_NOT_FOUND)))
+    (name-props (unwrap! 
+      (map-get? name-properties ((namespace namespace) (name name))) 
+      (err ERR_NAME_NOT_FOUND))))
+    (if (and (is-none (get registered-at name-props)) (is-some (get imported-at name-props)))
+      (ok false) ;; The name was imported and not launched - not subject to expiration, however should they expire if the namespace expire? - if so, todo(ludo)
+      (let (      ;; The name was registered
+        (registered-at (unwrap! 
+          (get registered-at name-props) (err ERR_PANIC)))
+        (lifetime 
+          (get renewal-rule namespace-props)))
+        (if (is-eq lifetime u0)
+          (ok false)
+          (ok (> block-height (+ lifetime registered-at))))))))
+
+;; todo(ludo): should be refactored - based on (is-name-lease-expired?)
+(define-read-only (is-name-in-grace-period? (namespace (buff 19)) (name (buff 16)))
+  (let (
+    (namespace-props (unwrap! 
+      (map-get? namespaces ((namespace namespace))) 
+      (err ERR_NAMESPACE_NOT_FOUND)))
+    (name-props (unwrap! 
+      (map-get? name-properties ((namespace namespace) (name name))) 
+      (err ERR_NAME_NOT_FOUND))))
+    (if (is-none (get registered-at name-props))
+      (ok false) ;; The name was imported - not subject to expiration
+      (let (      ;; The name was registered
+        (registered-at (unwrap! 
+          (get registered-at name-props) (err ERR_PANIC)))
+        (lifetime 
+          (get renewal-rule namespace-props)))
+        (if (is-eq lifetime u0)
+          (ok false)
+          (ok (and 
+            (> block-height (+ lifetime registered-at)) 
```

Comment:
Shouldn't this be `>=`?

Reply:
I guess it depends on how we have been / how will be documenting `lifetime`.
If I'm sending a TX, registering a name `ludo.id` for a lifetime of 1, and that the TX is being mined at block 10, then I think I would expect my name to be valid at block 11, but expired at block 12, otherwise a lifetime of 1 would be unusable?

Reply:
Ah, makes sense.  I thought when I was reading this that there could be a 1-gap block where a name could be expired, but not yet in the grace period.